### ISSUE TITLE
"java.lang.OutOfMemoryError: GC overhead limit exceeded" when indexing complex entities

### ISSUE THREAD

**Issue by geraldhumphries** [geraldhumphries/generator-jhipster-elasticsearch-reindexer#17]
This happens when indexing a large list of complex entities, maybe with a chain of eagerly loaded relationships.

What I think is happening: The GC never has a chance to clean up already-indexed objects because elasticsearchRepository is working with the whole list. Jackson initializes any lazily-loaded fields on the entities one-by-one growing the memory usage as time goes on.

Using pagination for this is probably not an option because the generated `findAllWithEagerRelationships` does not use pagination.

Maybe split the results up into subLists and index the subLists, to allow the GC to do its job?

**Comment by geraldhumphries** [comment]
Second solution seems to work for my dataset, but probably still has issues with a big enough list.

**Comment by FireWolf2007** [comment]
If allowed to skip many-to-many associations may help such code:
```
...
                Pageable pageable = new PageRequest(0, 500, null, "id");
                Page<T> list = jpaRepository.findAll(pageable);
                while (list.getNumberOfElements() > 0) {
                    elasticsearchRepository.save(list);
                    pageable = pageable.next();
                    list = jpaRepository.findAll(pageable);
                    // may need cache cleanup, see below
                }
...
```

With cache cleanup it allow stay in normal heap limit while processing large number of objects.
```
                        for (String cacheName : cacheManager.getCacheNames()) {
                            cacheManager.getCache(cacheName).clear();
                        }
```

**Comment by geraldhumphries** [comment]
Unfortunately the indexing needs to match the format when saved through the API, which includes many-to-many associations.

I will need to test cache cleanups to see if it helps.

**Comment by FireWolf2007** [comment]
I'm just realized that call to reindexForClass will work without transaction because it's call from same bean and transaction annotation is ignored.

**Comment by FireWolf2007** [comment]
After replacing reindexForClass call to newService.reindexForClass reindexing speed slow down from 6000obj/min to 1000obj/min. After full reindexing I will report about results :)

**Comment by FireWolf2007** [comment]
Example of ElasticsearchEntityIndexerService:
```
package ru.inquarta.telestat.service;

import org.hibernate.Hibernate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.CacheManager;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.elasticsearch.repository.ElasticsearchRepository;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import ru.inquarta.telestat.domain.*;

import java.io.Serializable;
import java.util.concurrent.atomic.AtomicInteger;

@Service
public class ElasticsearchEntityIndexerService {

    private final Logger log = LoggerFactory.getLogger(ElasticsearchEntityIndexerService.class);

    private final CacheManager cacheManager;

    private int cleanupCounter = 0;

    public ElasticsearchEntityIndexerService(CacheManager cacheManager) {
        this.cacheManager = cacheManager;
    }

    @Transactional(readOnly = true)
    public <T, ID extends Serializable> void reindexForClassByPages(Class<T> entityClass, JpaRepository<T, ID> jpaRepository,
                                                              ElasticsearchRepository<T, ID> elasticsearchRepository,
                                                              long totalCount) {
        String simpleName = entityClass.getSimpleName();
        Pageable pageable = new PageRequest(0, 500, null, "id");
        Page<T> list = jpaRepository.findAll(pageable);
        AtomicInteger savedCount = new AtomicInteger(0);
        while (list.getNumberOfElements() > 0) {
            log.trace("{}, list size: {}, saved {} of {}", simpleName, list.getNumberOfElements(), savedCount.get(), totalCount);

            if ("Entity1".equals(simpleName)) {
                for (T t : list) {
                    Entity1 p = (Entity1)t;
                    if (p.getRelation1() != null) {
                        Hibernate.initialize(p.getEntity2().getManyToManuyRelation1());
                    }
                }
            } else if ("Entity2".equals(simpleName)) {
                for (T t : list) {
                    Entity2 p = (Entity2)t;
                    Hibernate.initialize(p.getManyToManuyRelation1());
                }
            } else if ("Entity3".equals(simpleName)) {
                for (T t : list) {
                    Entity3 p = (Entity3)t;
                    Hibernate.initialize(p.getManyToManuyRelation3());
                }
            }

            elasticsearchRepository.save(list).forEach(a -> savedCount.incrementAndGet());
            clearCache();
            pageable = pageable.next();
            list = jpaRepository.findAll(pageable);
        }
        log.info("Elasticsearch: Indexed {} of {} rows for {} with findAll", savedCount.get(), totalCount, simpleName);
    }

    @Transactional(readOnly = true)
    public <T, ID extends Serializable> void reindexForClassByFindAll(JpaRepository<T, ID> jpaRepository,
                                                              ElasticsearchRepository<T, ID> elasticsearchRepository) {
        elasticsearchRepository.save(jpaRepository.findAll());
        clearCache();
    }

    private void clearCache() {
        if (cleanupCounter > 5) {
            log.trace("Elasticsearch: cleanup cache");
            for (String cacheName : cacheManager.getCacheNames()) {
                if (cacheName.startsWith("ru.inquarta.telestat.domain.")) {
                    cacheManager.getCache(cacheName).clear();
                    log.trace("Elasticsearch: cleanup cache {}", cacheName);
                } else {
                    log.trace("Elasticsearch: cleanup cache blocked {}", cacheName);
                }
            }
            cleanupCounter = 0;
        }
        cleanupCounter ++;
    }
}
```

I think logic with entities in reindexForClassByPages may be created while generation process.
But may be complex relations not covered in my example. I did not figured yet how it saved on reindex and on entity editing.

In ElasticsearchIndexService:

```
    private <T, ID extends Serializable> void reindexForClass(Class<T> entityClass, JpaRepository<T, ID> jpaRepository,
                                                              ElasticsearchRepository<T, ID> elasticsearchRepository) {
        elasticsearchTemplate.deleteIndex(entityClass);
        try {
            elasticsearchTemplate.createIndex(entityClass);
        } catch (IndexAlreadyExistsException e) {
            // Do nothing. Index was already concurrently recreated by some other service.
        }
        elasticsearchTemplate.putMapping(entityClass);
        long totalCount = jpaRepository.count();
        if (totalCount > 0) {
            try {
                elasticsearchEntityIndexerService.reindexForClassByPages(entityClass, jpaRepository, elasticsearchRepository, totalCount);
            } catch (Exception e) {
                log.error("Exception: {}", e.getMessage(), e);
                elasticsearchEntityIndexerService.reindexForClassByFindAll(jpaRepository, elasticsearchRepository);
                log.info("Elasticsearch: Indexed all rows for {}", entityClass.getSimpleName());
            }
        }
    }
```

**Comment by geraldhumphries** [comment]
I am going to consider this fixed after #38 so I am closing it.
